torchrl/data/rlhf/utils.py

```diff
@@ -0,0 +1,293 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Tuple
+
+import torch
+
+from tensordict import TensorDict
+from torch import Tensor
+from torch.nn import functional as F
+
+from torchrl.data.rlhf.prompt import PromptData
+from transformers import GenerationConfig
+
+
+class RolloutFromModel:
+    """A class for performing rollouts with language models.
+
+    Args:
+        model (transformers.Transformer): the model to be used. Should have a
+            :meth:`generate` method.
+        ref_model (transformers.Transformer): a frozen version of ``model``
+            where params are in their initial configuration.
+        reward_model: (nn.Module, tensordict.nn.TensorDictModule): a model which, given
+            input_ids and attention_mask, calculates rewards for each token and
+            end_scores (the reward for the final token in each sequence).
+        max_new_tokens (int, optional): the maximum length of the sequence.
+            Defaults to 50.
+    """
+
+    EOS_TOKEN_ID = 50256
+
+    def __init__(self, model, ref_model, reward_model, max_new_tokens=50):
+        self.model = model
+        self.ref_model = ref_model
+        self.reward_model = reward_model
+        self.max_new_tokens = max_new_tokens
+
+    def kl_step(self):
+        """Makes a step in the KL coefficient schedule."""
+        pass
+
+    @torch.no_grad()
+    def rollout_from_data(self, batch, kl_coef=0.1):
+        generated, log_probs, log_ratio = self.generate(batch)
+        return self.create_rollout_td(batch, generated, log_probs, log_ratio, kl_coef)
+
+    @torch.no_grad()
+    def create_rollout_td(self, batch, generated, log_probs, log_ratio, kl_coef=0.1):
+        """A TensorDict wrapper for generated data.
+
+        This function takes a batch plus the generated tokens and replicates the
+        tensordict structure that would have been obtained from a rollout with a TorchRL
+        env that sampled one token each timestep.
+
+        Args:
+            batch: A batch of data containing the original prompt together with a field
+                "rindex" indicating the right index of the prompt.
+            generated: The prompt together with generated tokens. This can be obtained
+                by calling the ``generate`` method.
+            log_probs: The log probabilities of the generated tokens. Can be obtained by
+                calling the ``generate`` method.
+            log_ratio: The log ratio of the probabilities of the generated tokens
+                according to the generative model and the reference model. Can be
+                obtained by calling the ``generate`` method.
+            kl_coef: Coefficient with which to multiply the KL term before subtracting
+                from the reward.
+        """
```

Comment:
Should we also describe the output structure? (e.g. to define what reward_raw, kl, etc. are?

Reply:
Yes, good idea, I'll add this